---
title: "1-DataAnalysis"
author: "Arlie McCarthy"
date: "02/05/2021"
output: github_document
---
# Code Setup and Data Import - because that is super-important

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(vegan)
library(tidyverse)
library(readr)
library(here)
library(janitor)
library(cowplot)
library(reshape2)
library(ggdendro)
library(dendextend)
library(scales)
library(worrms)
```

I have one data file with the count data from lab processing. The file needs some wrangling to calculate a density (counts per decimeter squared, aka 100cm2) for each quadrat sampled.

```{r calculating density}
taxon_count_data <- read_csv(here::here("data", "taxon_count_data_internal_pipework.csv"))
taxon_density <- taxon_count_data %>% 
  clean_names() %>% 
  mutate(id = paste(area, quadrat_number, sep = "_")) %>% #some quadrats had multiple 'sample' so changed to just area and quadrat
  mutate(count = case_when(taxon_grouping_qualifier == "colony" ~ count * 4,
                           taxon_grouping_qualifier == "fragment" ~ count * 0.25,
                           taxon_grouping_qualifier == "squares" ~ count * 4,
                           taxon_grouping_qualifier == "small colony" ~ count * 4,
                           taxon_grouping_qualifier == "large colony" ~ count * 8,
                           TRUE ~ count)) %>%  #converts colonial organisms to cm2, i.e. a count of 1 = 1cm2, density is then e.g. 1 cm2 per 1dm2
  mutate(taxon_verified = case_when(taxon == "mussel" ~ "mytilus", 
                                    taxon == "acorn barnacle" ~ "amphibalanus_improvisus",
                                    taxon == "barnacle type 2" ~ "amphibalanus_amphitrite",
                                    taxon == "clam" ~ "hiatellidae",
                                    taxon == "c.f. Jassa" ~ "jassa_marmorata", 
                                    taxon == "acorn barnacle ribbed" ~ "austrominius_modestus",
                                    taxon == "crab" ~ "carcinus_maenas",
                                    taxon == "encrusting bryozoan" ~ "membranipora_membranacea",
                                    taxon == "white tufted bryozoan" | taxon == "type 2 white tufted bryozoan" ~ "tricellaria_inopinata",
                                    taxon == "round amphipod" | taxon == "green eye amphipod" ~ "stenothoidae",
                                    taxon == "c.f. Conchoderma" ~ "conchoderma_auritum", 
                                    taxon == "brown bryozoan" ~ "bugula_neritina",
                                    taxon == "c.f. Corophium" ~ "monocorophium_sextonae",
                                    taxon == "caprellid" ~ "caprella_mutica",
                                    taxon == "c.f. Schizoporella" ~ "cryptosula_pallasiana",
                                    taxon == "spiked encrusting bryo" ~ "electra_pilosa",
                                    taxon == "mystery tunicates? large anemone?" ~ "Tunicata indet 1",
                                    taxon == "c.f. Caprella equilibra" ~ "caprella_equilibra",
                                    taxon == "anemone" ~ "anthozoa",
                                    taxon == "crab juvenile" ~ "Brachyura juvenile",
                                    taxon == "Hyas sp" ~ "hyas",
                                    taxon == "hydroid" ~ "hydroidolina",
                                    taxon == "tunicate" ~ "Tunicata indet 1",
                                    taxon == "polychaete" ~ "polychaeta",
                                    taxon == "gastropod" ~ "gastropoda",
                                    taxon == "juvenile barnacle" ~ "Balanoidea juvenile",
                                    taxon == "mystery thing" ~ "Tunicata indet 3",
                                    taxon == "pycnogonid" ~ "pycnogonidae",
                                    taxon == "other bivalve" ~ "cardiidae",
                                    taxon == "sponge" ~ "porifera",
                                    taxon == "squat lobster juvenile" ~ "Anomura juvenile",
                                    taxon == "red-eye amphipod" ~ "Amphipoda indet 1",
                                    taxon == "purple amphipod" ~ "Amphipoda indet 2",
                                    taxon == "clear_tunicates" ~ "Tunicata indet 1",
                                    taxon == "colonial tunicate" ~ "Tunicata indet 2",
                                    taxon == "crustacean larvae" ~ "Crustacea larvae indet",
                                    taxon == "conopeum reticulum" ~ "conopeum_reticulum",
                                    TRUE ~ taxon
                                    ))  %>% 
  mutate(full_taxon_group = paste(taxon_verified, taxon_grouping_qualifier, sep = "_")) %>% 
  filter(full_taxon_group != "mytilus sp_valves intact, no tissue",
         full_taxon_group != "NA_NA",
         full_taxon_group != "mussel sp_valves intact",
         full_taxon_group != "insect_NA",
         full_taxon_group != "crab exoskeleton_NA",
         taxon_verified != "polychaete tube")%>% 
  group_by(id, taxon_verified, taxon_grouping_qualifier, area, stage, quadrat_number) %>% 
  mutate(density = count*multiply_by/divide_by/area_sampled*100) %>% #counts per dm2
  summarise(avg_density = mean(density))
site_info <- taxon_density %>% 
  ungroup() %>% 
  select(id, area, stage, quadrat_number) %>% 
  group_by(id) %>% 
  slice(1)
```
Check the species names and allocate higher taxonomic groups
```{r}
taxa_list <- unique(taxon_density$taxon_verified) %>% 
  as_tibble() %>% 
  rename(taxon_name = value)
aphia_id <- wm_name2id_(name = taxa_list$taxon_name)
aphia_id_df <- aphia_id %>% 
  as.data.frame() %>% 
  pivot_longer(cols = everything(),
               names_to = "taxon_name",
               values_to = "id") 
classification <- wm_classification_(aphia_id) %>% 
  pivot_wider(id_cols = id, names_from = c(rank), values_from = c(scientificname), values_fill = NA) %>%
  mutate(id = as.integer(id))
taxa_list_full <- aphia_id_df %>% 
  left_join(classification, by = "id") %>% 
  right_join(taxa_list, by = "taxon_name")
plotting_names <- taxon_density %>% 
  rename(taxon_name = taxon_verified) %>% 
  left_join(taxa_list_full %>% rename(aphia_id = id), by = "taxon_name") %>% 
  mutate(plotting_name = case_when(!is.na(Species) ~ Species,
                                   is.na(Species) ~ Genus,
                                   is.na(Genus) ~ Family,
                                   is.na(Family) ~ Order,
                                   is.na(Order) ~ Class,
                                   is.na(Class) ~ Subphylum,
                                   is.na(Subphylum) ~ Phylum)) %>% 
  mutate(plotting_name = case_when(is.na(plotting_name) ~ Family,
                                   TRUE ~ plotting_name)) %>% 
  mutate(plotting_name = case_when(is.na(plotting_name) ~ Order,
                                   TRUE ~ plotting_name)) %>%
  mutate(plotting_name = case_when(is.na(plotting_name) ~ Class,
                                   TRUE ~ plotting_name)) %>%
  mutate(plotting_name = case_when(is.na(plotting_name) ~ Subphylum,
                                   TRUE ~ plotting_name)) %>%
  mutate(plotting_name = case_when(is.na(plotting_name) ~ Phylum,
                                   TRUE ~ plotting_name)) %>% 
  mutate(plotting_name = case_when(taxon_name == "Amphipoda indet 1" ~ "Amphipoda indet 1",
                                   taxon_name == "Amphipoda indet 2" ~ "Amphipoda indet 2",
                                   taxon_name == "Tunicata indet 1" ~ "Tunicata indet 1",
                                   taxon_name == "Tunicata indet 2" ~ "Tunicata indet 2",
                                   taxon_name == "Tunicata indet 3" ~ "Tunicata indet 3",
                                   taxon_name == "Crustacea larvae indet" ~ "Crustacea larvae indet",
                                   taxon_name == "Brachyura juvenile" ~ "Brachyura juvenile indet",
                                   taxon_name == "Balanoidea juvenile" ~ "Balanoidea juvenile indet",
                                   taxon_name == "Anomura juvenile" ~ "Anomura juvenile indet",
                                   TRUE ~ plotting_name)) %>%
  mutate(Phylum = case_when(taxon_name == "Amphipoda indet 1" ~ "Arthropoda",
                            taxon_name == "Amphipoda indet 2" ~ "Arthropoda",
                            taxon_name == "Tunicata indet 1" ~ "Chordata",
                            taxon_name == "Tunicata indet 2" ~ "Chordata",
                            taxon_name == "Tunicata indet 3" ~ "Chordata",
                            taxon_name == "Crustacea larvae indet" ~ "Arthropoda",
                            taxon_name == "Brachyura juvenile" ~ "Arthropoda",
                            taxon_name == "Balanoidea juvenile" ~ "Arthropoda",
                            taxon_name == "Anomura juvenile" ~ "Arthropoda",
                            TRUE ~ Phylum)) %>% 
  select(taxon_name, plotting_name, Phylum, stage, avg_density)
```


# Preparing Data for ordination

I take my data from counts to a community matrix, with species as variables and rows as samples

```{r nMDS preparation}
community_matrix_rough <- plotting_names %>% 
  group_by(id, plotting_name) %>% 
  summarise(avg_density = sum(avg_density)) %>% 
  pivot_wider(names_from = plotting_name, values_from = avg_density, values_fill = 0) %>% 
  ungroup()
#community_matrix_rough[is.na(community_matrix_rough)] <- 0
community_matrix <- community_matrix_rough %>% 
  column_to_rownames(var = "id") %>% 
  filter_all(any_vars(. != 0))
write_csv(community_matrix %>% 
  rownames_to_column(var = "id"), here("data","community_matrix.csv"), col_names = TRUE)
write_csv(site_info, here("data","site_info.csv"), col_names = TRUE)
```

## nMDS

```{r nMDS}
internal_pipe_nmds <- metaMDS(community_matrix, distance = "bray", k = 2, trymax = 500)
internal_pipe_nmds
data_scores <- as.data.frame(scores(internal_pipe_nmds))  #Using the scores function from vegan to extract the site scores and convert to a data.frame
data_scores$site <- rownames(internal_pipe_nmds)  # create a column of site names, from the rownames of data.scores
data_scores <- data_scores %>% 
  rownames_to_column(var = "id") %>% 
  left_join(site_info)
data_scores$n_species <- specnumber(community_matrix, MARGIN = 1)
data_scores$shannon <- diversity(community_matrix, index = "shannon")
data_scores$simpson <- diversity(community_matrix, index = "simpson")
species_scores <- as.data.frame(scores(internal_pipe_nmds, "species"))  #Using the scores function from vegan to extract the species scores and convert to a data.frame
species_scores$species <- rownames(species_scores)

```

Plotting with ggplot
```{r}
nmds_plot <- ggplot() + 
  #geom_text(data=species_scores,aes(x=NMDS1,y=NMDS2,label=species),alpha=0.5) +  # add the species labels
  #geom_point(data=data_scores,aes(x=NMDS1,y=NMDS2,shape=as_factor(stage), colour= as_factor(area)),size=3) + # add the point markers
  geom_point(data=data_scores,aes(x=NMDS1,y=NMDS2,shape = as_factor(stage), colour = n_species),size=3) + # add the point markers
  #geom_text(data=data_scores,aes(x=NMDS1,y=NMDS2,label=area),size=6,vjust=0) +  # add the site labels
  coord_equal() +
  theme_bw() +
  scale_colour_viridis_c("Number of taxa") +
  scale_shape("Stage of Pipework", labels = c("1 - sea chest", "2 - sea strainer", "3 - mid ships", "4 - overboard discharge"))
nmds_plot

save_plot(here("outputs", "nmds_plot.png"), nmds_plot, base_width = 183, base_height = 100, units = "mm")
```
Checking the nMDS

```{r}
goodness(internal_pipe_nmds) # Produces a results of test statistics for goodness of fit for each point

stressplot(internal_pipe_nmds) # Produces a Shepards diagram
```


# Density of different groups

Here I will make a stacked bar chart to show mean density of different taxonomic groups at each stage of the pipework

```{r}
mean_density_stage <- plotting_names %>% 
  group_by(id, plotting_name, area, stage, Phylum) %>% 
  summarise(avg_density = sum(avg_density)) %>% #this ensures that there is only one number per taxa per sample
  ungroup() %>% 
  filter(!is.na(avg_density)) %>% 
  group_by(plotting_name, stage, Phylum) %>% 
  summarise(avg_density = mean(avg_density))
```


Now to plot it. This plot will be aligned with a schematic diagram representing the pipework system, so that readers can visualise passage through the pipes and the changes in taxa over space. Ultimately, this will have to be grouped/coloured by more manageable chunks than all the taxa I have listed. But even without that, it's clear that the Jassa amphipods dominate right the way through. It's also clear that there isn't a great deal surviving in the pipework areas sampled in the middle of the ship, but they do survive transit and settle and grow in the final stage before overboard discharge. 

```{r}
pal <- c("#332288", "#117733", "#44AA99", "#88CCEE", "#DDCC77", "#CC6677", "#AA4499", "#882255")
stacked_bar_plot <- ggplot() +
  geom_bar(data = mean_density_stage,
           aes(x = stage, y = avg_density, fill = Phylum),
           colour = "white",
           stat = "identity",
           position = "stack") +
#  geom_line(data = mean_density_stage, #problem with line is that without 0 for stages where taxa are absent, it connects the line making it look like the density goes consistently up/down rather than sharp changes
#           aes(x = stage, y = avg_density, colour = taxon),
#           stat = "identity",
#           position = "stack") +
  labs(x = "Stage of Pipework") +
  ylab(expression(Mean~density~(individuals~per~dm^2))) +
  theme_bw() +
  scale_fill_manual("Phylum", values = pal)
  #scale_fill_viridis_d(option = "B", direction = 1, begin = 0.1)
stacked_bar_plot
save_plot(here("outputs", "stacked_bar_plot.png"), stacked_bar_plot, base_width = 183, base_height = 100, units = "mm")
```
numbers for richness in each stage

```{r}
mean_density_stage %>% 
  group_by(stage) %>% 
  summarise(count = n_distinct(plotting_name))
```

# Creating a heatmap with dendrograms

```{r}
species_names <- colnames(community_matrix)
# Obtain the dendrogram
dend <- as.dendrogram(hclust(vegdist(community_matrix, method = "bray")))
dend_data <- dendro_data(dend)
```

#### Setup the data, so that the layout is inverted (this is more "clear" than simply using coord_flip())
```{r}
segment_data <- with(
  segment(dend_data), 
  data.frame(x = y, y = x, xend = yend, yend = xend))
# Extract dendrogram terminal segment data
dendrogram_ends <- with(segment(dend_data) %>%
  filter(yend == 0),
  data.frame(x = y, y = x, xend = yend, yend = xend))
```

Use the dendrogram label data to position the sample labels

```{r}
sample_pos_table <- with(
  dend_data$labels, 
  data.frame(y_center = x, id = as.character(label), height = 1)) %>% 
  left_join(site_info, by = "id")
# bind the sample information to the dendrogram ends for plotting
dendrogram_ends <- bind_cols(dendrogram_ends, sample_pos_table)
```

Creating custom colours for the labels of stages

```{r}
axiscolour <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
  #brewer_pal(type = "seq", palette = "Dark2")(6)

sample_pos_table <- sample_pos_table %>% 
  mutate(labs_colour = case_when(stage == 1 ~ axiscolour[8],
                                 stage == 2 ~ axiscolour[2],
                                 stage == 3 ~ axiscolour[3],
                                 stage == 4 ~ axiscolour[4]))
dend_palette <- axiscolour[c(8, 2, 3, 4)]
```

Table to position the samples

I also prepare suitable names for plotting the taxa and colouring by phylum (using the same colour scheme as the stacked bar chart)
```{r}
species_pos_table <- data.frame("taxon" = species_names) %>%
  left_join(mean_density_stage %>% 
              rename("taxon" = plotting_name) %>% 
              group_by(taxon, Phylum) %>% 
              mutate(density = sum(avg_density)) %>% 
              select(-stage, -avg_density) %>% 
              slice(1) %>% 
              arrange(desc(density)), 
            by = "taxon") %>% 
  arrange(desc(density)) %>% 
  dplyr::mutate(x_center = (1:n()), 
         width = 1) %>%
  mutate(labs_colour = case_when(Phylum == "Annelida" ~ pal[1],
                                 Phylum == "Arthropoda" ~ pal[2],
                                 Phylum == "Bryozoa" ~ pal[3],
                                 Phylum == "Chordata" ~ pal[4],
                                 Phylum == "Cnidaria" ~ pal[5],
                                 Phylum == "Mollusca" ~ pal[6],
                                 Phylum == "Nemertea" ~ pal[7],
                                 Phylum == "Porifera" ~ pal[8])) %>% 
  select(-density)
# Neglecting the gap parameters
heatmap_data <- community_matrix %>% 
  mutate("id" = rownames(community_matrix)) %>% 
  pivot_longer(cols = !id, names_to = "taxon", values_to = "density")%>%
  left_join(sample_pos_table) %>%
  left_join(species_pos_table, by = "taxon")

# changing 0 to NA so that samples/samples with low density appear different to those with none.
heatmap_data[heatmap_data == 0] <- NA

# Limits for the vertical axes
sample_axis_limits <- with(
  sample_pos_table, 
  c(min(y_center - 0.5 * height), max(y_center + 0.5 * height))
) + 
  0.1 * c(-1, 1) # extra spacing: 0.1
```
#For the heatmap plot I have changed the 'fill' to log density rather than density, as I think that
#with raw densities only Jassa was anything other than blue really - this way gives a bit more idea of the
#variation. Easy to go back though if you don't like it!
```{r}
# Heatmap plot
plt_hmap <- ggplot(heatmap_data, 
                   aes(x = x_center, y = y_center, fill = log10(density), 
                       height = height, width = width)) + 
  geom_tile() +
  scale_fill_viridis_c("Log density", option = "cividis", na.value = "white", begin = 0) +
  scale_x_continuous(breaks = species_pos_table$x_center, 
                     labels = species_pos_table$taxon,
                     expand = c(0, 0)
                     ) + 
  # For the y axis, alternatively set the labels as: sample_position_table$gene
  scale_y_continuous(breaks = sample_pos_table[, "y_center"], 
                     labels = rep("", nrow(sample_pos_table)),
                     limits = sample_axis_limits,
                     expand = c(0, 0)
                     ) + 
  labs(x = "Taxa", y = "") +
  theme_bw() +
  theme(axis.text.x = element_text(color=species_pos_table$labs_colour, size = rel(0.8), hjust = 1, angle = 45), 
        # margin: top, right, bottom, and left
        plot.margin = unit(c(1, 0.2, 0.2, -0.7), "cm"), 
        panel.grid.minor = element_blank(),
        legend.position = "right",
        legend.direction = "vertical")
plt_hmap
```

```{r}
# Dendrogram plot
plt_dendr <- ggplot(segment_data) + 
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend)) + 
  geom_point(data = dendrogram_ends, aes(x= xend, y = yend, shape = as_factor(sample_pos_table$stage))) +
  scale_x_reverse() + 
  scale_y_continuous(breaks = sample_pos_table$y_center, 
                     labels = sample_pos_table$id, 
                     limits = sample_axis_limits, 
                     expand = c(0, 0)) + 
  labs(x = "Distance", y = "Sample", size = "") +
  scale_shape("Stage") +
  theme_bw() + 
  theme(panel.grid.minor = element_blank()) +
  theme(#axis.text.y = element_text(face = "bold"),
        legend.position = "left",
        legend.direction = "vertical")
plt_dendr
```

Plot them both together and save
```{r}
heat_dend_plot <- plot_grid(plt_dendr, plt_hmap, align = 'h', rel_widths = c(1, 1.5))
heat_dend_plot
save_plot(here("outputs", "heat_dend_plot.png"), heat_dend_plot, base_width = 183, base_height = 140, units = "mm")
```

# Pulling out key numbers

Highest densities
```{r}
max(plotting_names$avg_density, na.rm = TRUE)
plotting_names %>% 
  group_by(id, plotting_name, area, stage) %>% 
  summarise(avg_density = sum(avg_density)) %>%
  arrange(desc(avg_density)) %>%
  head()


plotting_names %>% 
  group_by(id, plotting_name, area, stage) %>% 
  summarise(avg_density = sum(avg_density)) %>%
  arrange(desc(avg_density)) %>%
  filter(stage == 4)
```